<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px"> 

## Lab: SQL Practice

_Author: Matt Brems_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [1]:
# Import pandas and create_engine from `sqlalchemy`

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data
engine = \
create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [2]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT *
FROM stores
"""

stores = pd.read_sql_query(sql, engine)
stores

store                         name store_status  \
0      2106  Hillstreet News and Tobacco            A   
1      2112               Mike's Liquors            I   
2      2113               Jamboree Foods            A   
3      2119           Manly Liquor Store            I   
4      2130         Sycamore Convenience            A   
...     ...                          ...          ...   
1968   9917       Honey Creek Distillery            A   
1969   9918  Paradise Distilling Company            A   
1970   9919            Dehner Distillery            A   
1971   9920        Louisiana Spirits LLC            A   
1972   9921                  maDIKwe USA            A   

                                          store_address address_info  
0     2217 College\nCedar Falls, IA 506130000\n(42.5...         None  
1     407 Sharp St.\nGlenwood, IA 515340000\n(41.046...         None  
2     1119 Market St. Box 71\nGowrie, IA 505430000\n...         None  
3     133 East Main\nManly, IA 504560000\n(43.286863...         None  
4     617 Sycamore\nWaterloo, IA 507030000\n(42.4978...         None  
...                                                 ...          ...  
1968  29538, 210th St.\nKeosauqua, IA 52565\n(40.740...         None  
1969  245, Railroad Ave\nDubuque, IA 52001\n(42.4873...         None  
1970  7500, University Ave\nClive, IA\n(41.600352803...         None  
1971  20909, South I-10 Frontage Rd\nLacassine, IA 7...         None  
1972  3101, Ingersoll Ave\nDes Moines, IA 50312\n(41...         None  

[1973 rows x 5 columns]

In [3]:
stores.shape[0]

1973

**Prompt 1 Answer**: 1973 rows

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [4]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT name, store
FROM stores
ORDER BY name
"""

pd.read_sql_query(sql, engine)

name  store
0     218 Fuel Express & Chubby's Liquor   4986
1             3047 Cub Foods / Iowa City   3047
2                  3060 Cub Foods / Ames   3060
3            3061 Cub Foods / Sioux City   3061
4                     3rd St Convenience   4845
...                                  ...    ...
1968       Xo Food And Liquor / Waterloo   3992
1969                       Xpress Liquor   4066
1970                      Yr Dollar Stop   4886
1971                Zapf's Pronto Market   4908
1972                      Z's Quickbreak   4874

[1973 rows x 2 columns]

**Prompt 2 Answer**: 3061 Cub Foods / Sioux City

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [13]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT vendor, category_name, btl_price, bottle_qty
FROM sales
WHERE category_name LIKE '%%IMPORTED VODKA%%'
ORDER BY 1
LIMIT 3
"""

# sql = """
# SELECT DISTINCT category_name
# FROM sales
# """
pd.read_sql_query(sql, engine)

vendor   category_name btl_price  bottle_qty
0  Adamba Imports Int'l Inc.  IMPORTED VODKA     $8.99          12
1      A Hardy / U.S.A. Ltd.  IMPORTED VODKA    $28.50           1
2      A Hardy / U.S.A. Ltd.  IMPORTED VODKA    $28.50           6

**Prompt 3 Answer**: 
1.	Adamba Imports Int'l Inc.	Bottle price: $8.99, quantity: 12

2.	A Hardy / U.S.A. Ltd.		Bottle price: $28.50, quantity:	1

3.	A Hardy / U.S.A. Ltd.		Bottle price: $28.50, quantity:	6

### Prompt 4

What is the total amount of liquor sold in February 2015?

In [16]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT SUM(bottle_qty) AS total_amount_liquor
FROM sales
WHERE DATE_PART('year', date) = 2015 AND DATE_PART('month', date) = 2
"""
pd.read_sql_query(sql, engine)

total_amount_liquor
0              1591962

**Prompt 4 Answer**: 1,591,962 bottles

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [19]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT vendor, COUNT(*) AS total_transaction, 
SUM(bottle_qty) AS bottles_sold, 
SUM(total) AS total_revanue
FROM sales
WHERE category_name LIKE '%%IMPORTED VODKA%%'
GROUP BY vendor
ORDER BY 2 DESC
"""

pd.read_sql_query(sql, engine)

vendor  total_transaction  bottles_sold  \
0       Pernod Ricard USA/Austin Nichols              67190        570679   
1                        Diageo Americas              31672        281482   
2                    Bacardi U.S.A. Inc.              29054        244756   
3        Constellation Wine Company Inc.              18174        148942   
4                        Jim Beam Brands              15123        155472   
5                                Proximo              12546         76791   
6                            Stoli Group               7894         63936   
7                         Luxco-St Louis               6257         54397   
8                   Bacardi U.S.A., Inc.               3632         31269   
9                   Imperial Brands Inc.               2987         22463   
10                Moet Hennessy USA Inc.               2403         12474   
11      Constellation Wine Company, Inc.               2353         17158   
12                               MHW Ltd               1159          4493   
13      Western Spirits Beverage Co. LLC                923          4225   
14                   Wilson Daniels Ltd.                815          2396   
15                                 LUXCO                698          7650   
16                 Imperial Brands, Inc.                427          3055   
17            Russian Standard Vodka USA                415          2421   
18               Moet Hennessy USA, Inc.                294          1209   
19                      Quadro Group LLC                289          1054   
20              Brown-Forman Corporation                270          1850   
21           WILLIAM GRANT AND SONS INC.                232          1119   
22                      Klin Spirits LLC                218           784   
23           Russian Standard Vodka, USA                 71           360   
24                                  None                 62           368   
25          WILLIAM GRANT AND SONS, INC.                 56           252   
26                 A Hardy / U.S.A. Ltd.                 33           320   
27           W. J. Deutsch And Sons Ltd.                 24           768   
28                     Jinro America Inc                 21          1056   
29              Firestarter Spirits Inc.                 20            54   
30                             TY KU LLC                 17            87   
31          Park Street Imports LLC Voli                 12            84   
32                      Sazerac Co. Inc.                 12            84   
33  Shaw Ross International Importers LL                 10           516   
34                     Klin Spirits, LLC                  5            13   
35            The Patron Spirits Company                  4            54   
36                     Wine Intelligence                  3           360   
37      Stanley Stawski Distributing Co.                  2            24   
38        Charles Jacquin Et. Cie., Inc.                  2            12   
39                     Phenix Brands LLC                  2            12   
40                 Ti Beverage Group Ltd                  2            24   
41                A Hardy / U.S.A., Ltd.                  1             5   
42             Adamba Imports Int'l Inc.                  1            12   

    total_revanue  
0     10902960.40  
1      5699011.89  
2      6006840.37  
3      2303316.27  
4      2468942.97  
5      1209047.17  
6      1445772.69  
7       625676.84  
8       793919.51  
9       277662.11  
10      361927.38  
11      252305.32  
12      109190.48  
13       51070.24  
14       84246.72  
15       94569.30  
16       37958.75  
17       40361.58  
18       35337.74  
19       25928.40  
20       25829.08  
21       20981.25  
22       15483.84  
23        5400.00  
24        6009.44  
25        4725.00  
26        9120.00  
27        8593.92  
28        7319.04  
29         994.14  
30        2429.85  
31   

**Prompt 5 Answer**: Pernod Ricard USA/Austin Nichols

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [20]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT date, vendor, btl_price - state_btl_cost AS Markup
FROM sales
"""

pd.read_sql_query(sql, engine)

date                            vendor  markup
0       2014-11-13  Pernod Ricard USA/Austin Nichols   $5.75
1       2014-11-13  Pernod Ricard USA/Austin Nichols   $5.75
2       2014-11-13  Pernod Ricard USA/Austin Nichols   $5.75
3       2014-11-13            Moet Hennessy USA Inc.   $8.58
4       2014-11-13                   Diageo Americas   $2.50
...            ...                               ...     ...
3049908 2014-11-13  Pernod Ricard USA/Austin Nichols   $5.75
3049909 2014-11-13  Pernod Ricard USA/Austin Nichols  $10.00
3049910 2014-11-13  Pernod Ricard USA/Austin Nichols   $5.75
3049911 2014-11-13  Pernod Ricard USA/Austin Nichols   $5.75
3049912 2014-11-13  Pernod Ricard USA/Austin Nichols  $10.00

[3049913 rows x 3 columns]

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [35]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT store, AVG(Markup/state_btl_cost)*100.0 AS Percentage_Markup
FROM (SELECT store, state_btl_cost, btl_price - state_btl_cost AS Markup
      FROM sales) AS temp
GROUP BY store
ORDER BY Percentage_Markup DESC
LIMIT 1
"""

pd.read_sql_query(sql, engine)

store  percentage_markup
0   4013          52.547124

**Prompt 7 Answer**: Store 4013 has highest average markup of 52.55% 

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [50]:
# Use a SQL query to find the answer to the above prompt.

sql = """
WITH temp3 AS (SELECT store, AVG(Percentage_Markup) AS average_markup, 
               AVG(min_price) AS min_markup, AVG(max_price) as max_markup
               FROM (SELECT store, Markup/state_btl_cost*100.0 AS Percentage_Markup,
               MIN(Markup/state_btl_cost*100.0) OVER (PARTITION BY store) AS min_price,
               MAX(Markup/state_btl_cost*100.0) OVER (PARTITION BY store) AS max_price
               FROM (SELECT store, state_btl_cost, btl_price - state_btl_cost AS Markup
                       FROM sales) AS temp) AS temp2
               GROUP BY store
               ORDER BY average_markup DESC
               LIMIT 5)
SELECT t3.store, name, store_address, average_markup, min_markup, max_markup
FROM temp3 AS t3
LEFT JOIN stores as s
ON s.store = t3.store

"""

pd.read_sql_query(sql, engine)

store                        name  \
0   4013      Wal-Mart 0841 / Tipton   
1   4024  Wal-Mart 1546 / Iowa Falls   
2   4266  Wal-Mart 1683 / Shenandoah   
3   3833    Wal-Mart 3394 / Atlantic   
4   3660   Wal-Mart 2935 / Knoxville   

                                       store_address  average_markup  \
0  1126 Highway 38 North\nTipton, IA 527720000\n(...       52.547124   
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...       52.149880   
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...       51.890342   
3  1905 East 7th St\nAtlantic, IA 500220000\n(41....       51.509378   
4  814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...       51.459084   

   min_markup  max_markup  
0   49.857143  670.876289  
1   49.846154  670.876289  
2   49.846154  670.876289  
3   49.846154  670.876289  
4   49.846154  670.876289

**Prompt 8 Answer**:

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery).

In [82]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT *
FROM (SELECT s.county, SUM(bottle_qty::FLOAT/population) AS n_bottle,
ROUND(AVG(bottle_size),1) AS bottle_size
FROM sales as s
JOIN counties as c
ON c.county = s.county
JOIN products as p
ON s.item = p.item_no
GROUP BY s.county
ORDER BY 2 DESC
LIMIT 9) AS temp
ORDER BY 3 DESC
"""

highest_bottle = pd.read_sql_query(sql, engine)
highest_bottle

county   n_bottle  bottle_size
0        Kossuth  11.373480       1041.0
1    Cerro Gordo  14.121062        986.4
2      Dickinson  22.256315        974.1
3        Johnson  13.164339        895.4
4  Pottawattamie  11.578877        893.7
5           Polk  15.081873        892.6
6           Linn  12.832024        886.3
7          Scott  13.901818        871.6
8     Black Hawk  15.385132        862.9

**Prompt 9 Answer**:

In [83]:
# Which nine counties sell the highest
# number of bottles of liquor per capita (per person)? 
list(highest_bottle['county'])

['Kossuth',
 'Cerro Gordo',
 'Dickinson',
 'Johnson',
 'Pottawattamie',
 'Polk',
 'Linn',
 'Scott',
 'Black Hawk']

In [70]:
# which county sells (on average) the largest bottle size?
# Kossuth: 1041.0

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [76]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(*)
FROM (SELECT btl_price*bottle_qty AS price_qty, total::MONEY
     FROM sales) AS temp
WHERE total != price_qty
"""

pd.read_sql_query(sql, engine)

count
0      0

**Prompt 10 Answer**: 0

#### Additional question
1. How many total products are in the Products table?

2. Who are the top most diverse vendors (i.e. they have the highest number of distinct products)? How many different products do they have?

3. Which products sell the best by total number of unit sales? (Hint: you will need to find a way of doing a cross-join between the sales and product tables.)

4. Which products sell the best by total dollar value of sales?

5. What are the top 10 categories of liquor sold based on the total amount of sales revenue?

6. Which rum products have sales greater than $10,000? How about whiskey or vodka products?

7. Which county sold the most amount of vodka during February 2014?

8. Which counties were in the top 10 counties for vodka sales in any month in 2014?

9. Create a report that shows how many times a county appeared in the “top 10 counties for vodka sales in a month” list over the course of 2014.

10. What is the trend of sales by month? Break up variables such as bottle_price or liter_size into categories (for example: cheap, medium, or expensive).

11. Which stores sell one of the top five most expensive bottles of alcohol?

12. How many stores have more than $2,000,000 in total sales?

13. How many stores have an average bottle price greater than $20?

14. Which stores have the highest sales of items over 90 proof? (edited) 

In [84]:
# 1. How many total products are in the Products table?
sql ="""
SELECT COUNT(item_no)
FROM products
"""
pd.read_sql_query(sql, engine)
# Ans: 9977

count
0   9977

In [89]:
# 2. Who are the top most diverse vendors (i.e. they have the highest number of distinct products)? 
# How many different products do they have?
sql = """
SELECT vendor_name, COUNT(DISTINCT item_no)
FROM products
GROUP BY vendor_name
ORDER BY 2 DESC
LIMIT 1
"""
pd.read_sql_query(sql, engine)
# Ans: Jim Beam Brands --> unique product: 925

vendor_name  count
0  Jim Beam Brands    925

In [3]:
# 3. Which products sell the best by total number of unit sales? 
# (Hint: you will need to find a way of doing a cross-join between the sales and product tables.)
sql = """
SELECT item_description, SUM(bottle_qty) AS total_bottle
FROM sales AS s
JOIN products AS p
ON s.item = p.item_no
GROUP BY 1
ORDER BY 2 DESC
LIMIT 1
"""
pd.read_sql_query(sql, engine)
# Ans: Black Velvet: 1,672,344 bottles

item_description  total_bottle
0     Black Velvet       1672344

In [4]:
# 4. Which products sell the best by total dollar value of sales?
sql = """
SELECT item_description, SUM(total) AS total_sales
FROM sales AS s
JOIN products AS p
ON s.item = p.item_no
GROUP BY 1
ORDER BY 2 DESC
LIMIT 1
"""
pd.read_sql_query(sql, engine)
# Ans: Black Velvet: 18,315,550 dollar

item_description  total_sales
0     Black Velvet  18315550.64

In [8]:
# 5. What are the top 10 categories of liquor sold based on the total amount of sales revenue?
sql = """
SELECT p.category_name, SUM(total) AS total_revenue
FROM sales AS s
JOIN products AS p
ON s.item = p.item_no
GROUP BY 1
ORDER BY 2 DESC
LIMIT 10
"""
pd.read_sql_query(sql, engine)

category_name  total_revenue
0                 CANADIAN WHISKIES    48059037.63
1                    80 PROOF VODKA    48040023.20
2                        SPICED RUM    31586392.99
3                    IMPORTED VODKA    23884662.13
4                           TEQUILA    21413525.94
5         STRAIGHT BOURBON WHISKIES    20904099.16
6                   WHISKEY LIQUEUR    19346811.30
7                TENNESSEE WHISKIES    17600801.58
8  PUERTO RICO & VIRGIN ISLANDS RUM    12729072.76
9                  BLENDED WHISKIES    12018198.55

In [12]:
# 6. Which rum products have sales greater than $10,000? How about whiskey or vodka products?
sql = """
SELECT p.category_name, SUM(total) AS total_revenue
FROM sales AS s
JOIN products AS p
ON s.item = p.item_no
WHERE p.category_name LIKE '%%RUM%%'
GROUP BY 1
HAVING SUM(total) > 10000
ORDER BY 1
"""
pd.read_sql_query(sql, engine)

category_name  total_revenue
0                      BARBADOS RUM      396024.49
1                      FLAVORED RUM     8011699.50
2                       JAMAICA RUM      371404.44
3  PUERTO RICO & VIRGIN ISLANDS RUM    12729072.76
4                        SPICED RUM    31586392.99

In [13]:
sql = """
SELECT p.category_name, SUM(total) AS total_revenue
FROM sales AS s
JOIN products AS p
ON s.item = p.item_no
WHERE p.category_name LIKE '%%WHISKEY%%' OR p.category_name LIKE '%%VODKA%%'
GROUP BY 1
HAVING SUM(total) > 10000
ORDER BY 1
"""
pd.read_sql_query(sql, engine)

category_name  total_revenue
0        100 PROOF VODKA      319636.81
1         80 PROOF VODKA    48040023.20
2         FLAVORED VODKA    11538501.82
3         IMPORTED VODKA    23884662.13
4  IMPORTED VODKA - MISC     9072915.58
5        LOW PROOF VODKA       75863.65
6      OTHER PROOF VODKA       56499.45
7        WHISKEY LIQUEUR    19346811.30

In [18]:
# 7. Which county sold the most amount of vodka during February 2014?
sql = """
SELECT s.county, SUM(bottle_qty) AS total_bottle
FROM sales AS s
JOIN products AS p
ON s.item = p.item_no
JOIN counties AS c
ON c.county = s.county
WHERE p.category_name LIKE '%%VODKA%%' AND DATE_PART('year',date) = 2014 \
AND DATE_PART('month', date) = 2
GROUP BY 1
ORDER BY 2 DESC
LIMIT 1
"""
pd.read_sql_query(sql, engine)
# Ans: Polk

county  total_bottle
0   Polk        202036

In [19]:
# 8. Which counties were in the top 10 counties for vodka sales in any month in 2014?
sql = """
SELECT s.county, SUM(total) AS total_sale
FROM sales AS s
JOIN products AS p
ON s.item = p.item_no
JOIN counties AS c
ON c.county = s.county
WHERE p.category_name LIKE '%%VODKA%%' AND DATE_PART('year',date) = 2014
GROUP BY 1
ORDER BY 2 DESC
LIMIT 10
"""
pd.read_sql_query(sql, engine)
# Ans: Polk, Linn, Scott, Johnson, Black Hawk, Pottawattamie, Dubuque
#      Story, Woodbury, Dallas

county   total_sale
0           Polk  20992773.49
1           Linn   7713951.02
2          Scott   6610080.92
3        Johnson   5871390.79
4     Black Hawk   4972849.31
5  Pottawattamie   2700297.72
6        Dubuque   2645905.67
7          Story   2605483.32
8       Woodbury   2407013.06
9         Dallas   1879411.18

In [71]:
# 9. Create a report that shows how many times a county appeared 
# in the “top 10 counties for vodka sales in a month” list over the course of 2014.
sql = """
WITH county_month AS (SELECT s.county, total, DATE_PART('month',date) AS Month
                      FROM sales AS s
                      JOIN products AS p
                      ON s.item = p.item_no
                      JOIN counties AS c
                      ON c.county = s.county
                      WHERE p.category_name LIKE '%%VODKA%%' AND DATE_PART('year',date) = 2014
                      ORDER BY Month)

SELECT county, COUNT(county) AS number_appear
FROM (SELECT county, month, SUM(total) AS total_sale,
                   DENSE_RANK() OVER (PARTITION BY month ORDER BY SUM(total) DESC) AS rank_month
                   FROM county_month
                   GROUP BY county, month) AS temp1
WHERE rank_month <= 10
GROUP BY county
ORDER BY 2 DESC
"""
top_10_appear_2014 = pd.read_sql_query(sql, engine)
top_10_appear_2014

county  number_appear
0            Polk             12
1            Linn             12
2           Scott             12
3      Black Hawk             12
4   Pottawattamie             12
5         Dubuque             12
6         Johnson             12
7        Woodbury             11
8           Story             10
9     Cerro Gordo              7
10         Dallas              6
11     Des Moines              1
12      Dickinson              1

In [51]:
# 10. What is the trend of sales by month? Break up variables such as bottle_price 
# or liter_size into categories (for example: cheap, medium, or expensive).
# btl_price: 0-100 -> cheap, 100-500 -> medium, 500+ -> expensive
sql = """
SELECT year, month, category_price, SUM(total)
FROM   (SELECT DATE_PART('year', date) AS year,
        DATE_PART('month',date) AS month, 
        CASE    WHEN btl_price <= 100::MONEY THEN 'cheap'
                WHEN btl_price > 100::MONEY AND btl_price <= 500::MONEY THEN 'medium'
                ELSE 'expensive' END AS category_price, total
        FROM sales) AS temp
GROUP BY 1, 2, 3
ORDER BY 1, 2, 3
"""
pd.read_sql_query(sql, engine)

year  month category_price          sum
0   2014.0    1.0          cheap  40290788.28
1   2014.0    1.0      expensive      1495.92
2   2014.0    1.0         medium     38739.76
3   2014.0    2.0          cheap  38472775.12
4   2014.0    2.0      expensive     28785.60
5   2014.0    2.0         medium     42027.70
6   2014.0    3.0          cheap  38497134.26
7   2014.0    3.0      expensive     19190.40
8   2014.0    3.0         medium     87219.16
9   2014.0    4.0          cheap  45312717.84
10  2014.0    4.0         medium     75776.80
11  2014.0    5.0          cheap  41994480.62
12  2014.0    5.0      expensive     17445.64
13  2014.0    5.0         medium     98142.38
14  2014.0    6.0          cheap  24362298.01
15  2014.0    6.0      expensive      3823.08
16  2014.0    6.0         medium     48476.08
17  2014.0    7.0          cheap  22056253.42
18  2014.0    7.0      expensive      8227.56
19  2014.0    7.0         medium     32255.95
20  2014.0    8.0          cheap  21183400.59
21  2014.0    8.0         medium     23947.25
22  2014.0    9.0          cheap  23403213.96
23  2014.0    9.0      expensive      3823.08
24  2014.0    9.0         medium     33922.89
25  2014.0   10.0          cheap  26641591.17
26  2014.0   10.0      expensive     19090.20
27  2014.0   10.0         medium     91920.83
28  2014.0   11.0          cheap  20177167.44
29  2014.0   11.0      expensive      4197.88
30  2014.0   11.0         medium     62845.22
31  2014.0   12.0          cheap   9273782.25
32  2014.0   12.0         medium     32440.61
33  2015.0    1.0          cheap  18525200.28
34  2015.0    1.0         medium     33075.83
35  2015.0    2.0          cheap  21255465.30
36  2015.0    2.0      expensive      5923.94
37  2015.0    2.0         medium     33961.31

In [41]:
# Using to check price 
# sql = """
# SELECT DISTINCT btl_price
# FROM sales
# ORDER BY 1 DESC
# LIMIT 100
# """
# pd.read_sql_query(sql, engine)

In [57]:
# 11. Which stores sell one of the top five most expensive bottles of alcohol?
sql = """
SELECT name, btl_price
FROM   (SELECT name, btl_price, 
        DENSE_RANK() OVER (ORDER BY btl_price DESC) AS rank_btl_price
        FROM sales AS s
        JOIN stores AS st
        ON st.store = s.store
        ORDER BY rank_btl_price) AS temp
WHERE rank_btl_price <= 5
"""
pd.read_sql_query(sql, engine)

name  btl_price
0   Hy-Vee Food and Drug #6 / Cedar Rapi  $8,700.00
1    Hy-vee Food Store #5 / Cedar Rapids  $8,700.00
2             Sam's Club 6514 / Waterloo  $2,398.80
3       Sam's Club 6472 / Council Bluffs  $2,398.80
4       Sam's Club 6472 / Council Bluffs  $2,398.80
5         Sam's Club 8162 / Cedar Rapids  $2,398.80
6         Sam's Club 8162 / Cedar Rapids  $2,398.80
7             Sam's Club 6514 / Waterloo  $2,398.80
8                 Sam's Club 6568 / Ames  $2,398.80
9           Sam's Club 6432 / Sioux City  $2,398.80
10           Sam's Club 8238 / Davenport  $2,398.80
11                Sam's Club 6568 / Ames  $2,398.80
12          Sam's Club 6432 / Sioux City  $2,398.80
13           Sam's Club 8238 / Davenport  $2,398.80
14             Sam's Club 4973 / Dubuque  $2,398.80
15     Sam's Club 6344 / Windsor Heights  $2,398.80
16     Sam's Club 6344 / Windsor Heights  $2,398.80
17             Sam's Club 4973 / Dubuque  $2,398.80
18                        Central City 2  $2,098.94
19             Grand Falls Casino Resort  $2,098.94
20             Grand Falls Casino Resort  $2,098.94
21           Charlie's Wine and Spirits,  $2,098.94
22                  Quicker Liquor Store    $845.10
23                        John's Grocery    $845.10
24    Hy-Vee Wine and Spirits / Waterloo    $747.96
25    Hy-Vee Wine and Spirits / Waterloo    $747.96
26                        John's Grocery    $747.96
27           Oelwein Bottle and Can Inc.    $747.96
28                    Hy-Vee / Urbandale    $747.96
29          Hy-vee #3 / Bdi / Des Moines    $747.96
30                     Benz Distributing    $747.96
31    Hy-Vee Wine and Spirits / Waterloo    $747.96
32             Central City Liquor, Inc.    $747.96

In [59]:
# 12. How many stores have more than $2,000,000 in total sales?
sql = """
SELECT COUNT(store)
FROM(SELECT s.store, SUM(total) AS total_sale
FROM sales AS s
JOIN stores AS st
ON st.store = s.store
GROUP BY s.store
HAVING SUM(total) > 2000000) AS temp
"""
pd.read_sql_query(sql, engine)
# Ans: 24

count
0     24

In [23]:
# 13. How many stores have an average bottle price greater than $20?
sql = """
SELECT COUNT(store)
FROM   (SELECT store, AVG(avg_btl_price) AS avg_btl_price
        FROM(SELECT store, item, AVG(btl_price::NUMERIC) AS avg_btl_price
             FROM sales 
             GROUP BY 1, 2) AS temp
        GROUP BY 1
        HAVING AVG(avg_btl_price) > 20) AS temp2
"""
pd.read_sql_query(sql, engine)

count
0     54

In [41]:
# 14. Which stores have the highest sales of items over 90 proof? (edited)
sql = """
SELECT t1.store, name, total_sale
FROM   (SELECT s.store, SUM(total) AS total_sale
        FROM sales AS s
        JOIN products AS p
        ON p.item_no = s.item
        WHERE proof::INT > 90
        GROUP BY 1
        ORDER BY 2 DESC
        LIMIT 1) AS t1
JOIN stores AS st
ON st.store = t1.store
"""
pd.read_sql_query(sql, engine)

store                          name  total_sale
0   2633  Hy-vee #3 / Bdi / Des Moines   689442.44

#### A nice time for a Window Function maybe?
- Store 2238 (Adventureland Inn at 3200 Adventureland Dr) sold \\$ 883.24 in April and \\$ 27,526.38 in May, for a 3017% growth rate. That was the highest percentage month-on-month growth rate. Create a query that shows this and the next 9 highest after that.

In [67]:
sql = """
SELECT year, month, total_sale, ROUND((total_sale - lag)/lag*100,2) AS monthly_growth_rate
FROM (SELECT year, month, total_sale, LAG(total_sale,1) OVER() AS lag
      FROM    (SELECT year, month, SUM(total) AS total_sale
               FROM   (SELECT store, DATE_PART('year',date) AS year,
                       DATE_PART('month', date) AS month, total
                       FROM sales
                       WHERE store = 2238) AS st2238
                GROUP BY 1, 2
                ORDER BY 1, 2) AS temp2) AS temp3
WHERE lag IS NOT NULL
"""
pd.read_sql_query(sql, engine)

year  month  total_sale  monthly_growth_rate
0   2014.0    2.0     2527.32                76.57
1   2014.0    3.0     1336.00               -47.14
2   2014.0    4.0      883.24               -33.89
3   2014.0    5.0    27526.38              3016.52
4   2014.0    6.0    25511.07                -7.32
5   2014.0    7.0    28972.18                13.57
6   2014.0    8.0    14254.21               -50.80
7   2014.0    9.0      503.89               -96.46
8   2014.0   10.0      297.12               -41.03
9   2014.0   11.0      347.51                16.96
10  2015.0    1.0     1963.67               465.07
11  2015.0    2.0      993.31               -49.42

#### Much more challenging. Pure Bonus level questions.
- We think the data might have been corrupted in some way. The category listed in the sales table doesn’t always match up with the category in the products table. How many times has this happened, and can you find any patterns to it?

In [79]:
sql = """
SELECT s.category_name AS sales_category, p.category_name AS product_category, description, item_description
FROM sales AS s
JOIN products AS p
ON p.item_no = s.item
WHERE s.category_name != p.category_name                
"""
category_diff = pd.read_sql_query(sql, engine)
category_diff
# Ans: 2020 times happened

sales_category                    product_category  \
0     DECANTERS & SPECIALTY PACKAGES  MISC. IMPORTED CORDIALS & LIQUEURS   
1     DECANTERS & SPECIALTY PACKAGES  MISC. IMPORTED CORDIALS & LIQUEURS   
2     DECANTERS & SPECIALTY PACKAGES  MISC. IMPORTED CORDIALS & LIQUEURS   
3     DECANTERS & SPECIALTY PACKAGES  MISC. IMPORTED CORDIALS & LIQUEURS   
4     DECANTERS & SPECIALTY PACKAGES  MISC. IMPORTED CORDIALS & LIQUEURS   
...                              ...                                 ...   
2015                      SPICED RUM      DECANTERS & SPECIALTY PACKAGES   
2016         IMPORTED GRAPE BRANDIES  MISC. IMPORTED CORDIALS & LIQUEURS   
2017         IMPORTED GRAPE BRANDIES  MISC. IMPORTED CORDIALS & LIQUEURS   
2018         IMPORTED GRAPE BRANDIES  MISC. IMPORTED CORDIALS & LIQUEURS   
2019         IMPORTED GRAPE BRANDIES  MISC. IMPORTED CORDIALS & LIQUEURS   

              description     item_description  
0      Jagermeister Spice   Jagermeister Spice  
1      Jagermeister Spice   Jagermeister Spice  
2      Jagermeister Spice   Jagermeister Spice  
3      Jagermeister Spice   Jagermeister Spice  
4      Jagermeister Spice   Jagermeister Spice  
...                   ...                  ...  
2015  Captain Morgan 1671  Captain Morgan 1671  
2016          D'usse VSOP          D'usse VSOP  
2017          D'usse VSOP          D'usse VSOP  
2018          D'usse VSOP          D'usse VSOP  
2019          D'usse VSOP          D'usse VSOP  

[2020 rows x 4 columns]

- The store_address field in the stores table actually contains three rows of text. Quite often the latitude and longitude are in the last line of text. Create a query that shows the geo-locatable stores in latitude order (i.e. show the stores from the most northerly to the most southerly).

In [124]:
sql = """
SELECT store, name,
REVERSE(SUBSTRING(rev_add, 2, POSITION('-' IN rev_add))) AS longitude ,
REVERSE(SUBSTRING(rev_add, POSITION(',' IN rev_add)+1,ABS(POSITION('(' IN rev_add) - POSITION(',' IN rev_add)-1))) AS latitude
FROM (SELECT store, name, REVERSE(store_address) AS rev_add
      FROM stores) AS temp
ORDER BY 3
"""
pd.read_sql_query(sql, engine)

store                                 name            longitude  \
0      2543       Hy-vee Food Store #1 / Ottumwa                        
1      4706           Walgreens #05144 / Clinton   -90.18207912099996   
2      4480             Circle K #6602 / Clinton   -90.18392888699998   
3      2806                  Osco #881 / Clinton   -90.18483592499996   
4      3759            Liquor Box, The / Clinton   -90.18991313699996   
...     ...                                  ...                  ...   
1968   4454           Kum & Go #248 / Sioux City   -96.47450478499997   
1969   3763       Main Street Liquors / Hawarden   -96.48527861999997   
1970   2155                Calliope Liquor Sales   -96.48798763499997   
1971   3798                  Old Calliope Liquor   -96.48798763499997   
1972   4525  Casey's General Store #2425 / Akron   -96.54957560199995   

                latitude  
0      incy Ave\nOttuwma  
1     41.866335062000076  
2      41.86147550800007  
3      41.85870901000004  
4     41.850745360000076  
...                  ...  
1968  42.514937316000044  
1969  42.996129050000036  
1970   43.00929963100003  
1971   43.00929963100003  
1972  42.834886589000064  

[1973 rows x 4 columns]

In [116]:
# Incomplete store address
sql= """
SELECT store, store_address
FROM (SELECT store, name, store_address
      FROM stores
      WHERE store = 2543) AS temp

"""
pd.read_sql_query(sql, engine)

store                                   store_address
0   2543  1025 North Quincy Ave\nOttuwma, IA 525010000\n

In [131]:
sql = """
SELECT store, name,
CASE WHEN POSITION(')'IN rev_add) = 1 THEN REVERSE(SUBSTRING(rev_add, 2, POSITION('-' IN rev_add))) ELSE '0' END AS longitude ,
CASE WHEN POSITION(')'IN rev_add) = 1 THEN \
     REVERSE(SUBSTRING(rev_add, POSITION(',' IN rev_add)+1,ABS(POSITION('(' IN rev_add) - POSITION(',' IN rev_add)-1))) ELSE '0' END AS latitude
FROM (SELECT store, name, REVERSE(store_address) AS rev_add
      FROM stores) AS temp
ORDER BY 3
"""
pd.read_sql_query(sql, engine)

store                                 name            longitude  \
0      2543       Hy-vee Food Store #1 / Ottumwa                    0   
1      4706           Walgreens #05144 / Clinton   -90.18207912099996   
2      4480             Circle K #6602 / Clinton   -90.18392888699998   
3      2806                  Osco #881 / Clinton   -90.18483592499996   
4      3759            Liquor Box, The / Clinton   -90.18991313699996   
...     ...                                  ...                  ...   
1968   4454           Kum & Go #248 / Sioux City   -96.47450478499997   
1969   3763       Main Street Liquors / Hawarden   -96.48527861999997   
1970   2155                Calliope Liquor Sales   -96.48798763499997   
1971   3798                  Old Calliope Liquor   -96.48798763499997   
1972   4525  Casey's General Store #2425 / Akron   -96.54957560199995   

                latitude  
0                      0  
1     41.866335062000076  
2      41.86147550800007  
3      41.85870901000004  
4     41.850745360000076  
...                  ...  
1968  42.514937316000044  
1969  42.996129050000036  
1970   43.00929963100003  
1971   43.00929963100003  
1972  42.834886589000064  

[1973 rows x 4 columns]